File: common/src/main/kotlin/net/impleri/playerskills/skills/specialized/SpecializedSkill.kt
package net.impleri.playerskills.skills.specialized

import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.TeamMode
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation

open class SpecializedSkill(
  name: ResourceLocation,
  value: String?,
  description: String?,
  options: List<String>?,
  changesAllowed: Int?,
  teamMode: TeamMode?,
  notify: Boolean?,
  notifyKey: String?,
) : Skill<String>(
  name,
  SpecializedSkillType.NAME,
  value,
  description,
  options ?: ArrayList(),
  changesAllowed ?: Skill.UNLIMITED_CHANGES,
  teamMode ?: TeamMode.off(),
  notify ?: false,
  notifyKey,
) {
  override fun copy(value: String?, changesAllowed: Int): Skill<String> {
    return SpecializedSkill(name, value, description, options, changesAllowed, teamMode, notify, notifyKey)
  }

  override fun getDefaultNotification(): Component {
    return Component.translatable(
      "playerskills.notify.specialized_skill_selected",
      formatSkillName(),
      formatSkillValue(),
    )
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/specialized/SpecializedSkill.scala
package net.impleri.playerskills.skills.specialized

import net.impleri.playerskills.api.skills.ChangeableSkillOps
import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.TeamMode
import net.minecraft.resources.ResourceLocation

case class SpecializedSkill(
  override val name: ResourceLocation,
  override val value: Option[String] = None,
  override val description: Option[String] = None,
  override val options: List[String] = List.empty,
  override val changesAllowed: Int = Skill.UNLIMITED_CHANGES,
  override val teamMode: TeamMode = TeamMode.Off(),
  override val announceChange: Boolean = false,
  override val notifyKey: Option[String] = None,
) extends Skill[String] with ChangeableSkillOps[String, SpecializedSkill] {
  override val skillType: ResourceLocation = SpecializedSkillType.NAME

  override protected[playerskills] def mutate(value: Option[String], changesAllowed: Int): SpecializedSkill = {
    copy(
      value = value, changesAllowed = changesAllowed,
    )
  }

  override def getMessageKey: String = "playerskills.notify.specialized_skill_selected"
}
